routing/payment_session.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// paymentSession is used during an HTLC routings session to prune the local
+// chain view in response to failures, and also report those failures back to
+// missionControl. The snapshot copied for this session will only ever grow,
+// and will now be pruned after a decay like the main view within mission
+// control. We do this as we want to avoid the case where we continually try a
+// bad edge or route multiple times in a session. This can lead to an infinite
+// loop if payment attempts take long enough. An additional set of edges can
+// also be provided to assist in reaching the payment's destination.
+type paymentSession struct {
+	pruneViewSnapshot graphPruneView
+
+	additionalEdges map[Vertex][]*channeldb.ChannelEdgePolicy
+
+	bandwidthHints map[uint64]lnwire.MilliSatoshi
+
+	// errFailedFeeChans is a map of the short channel IDs that were the
+	// source of policy related routing failures during this payment attempt.
+	// We'll use this map to prune out channels when the first error may not
+	// require pruning, but any subsequent ones do.
+	errFailedPolicyChans map[edgeLocator]struct{}
+
+	mc *missionControl
+
+	haveRoutes     bool
+	preBuiltRoutes []*Route
+}
+
+// ReportVertexFailure adds a vertex to the graph prune view after a client
+// reports a routing failure localized to the vertex. The time the vertex was
+// added is noted, as it'll be pruned from the shared view after a period of
+// vertexDecay. However, the vertex will remain pruned for the *local* session.
+// This ensures we don't retry this vertex during the payment attempt.
+func (p *paymentSession) ReportVertexFailure(v Vertex) {
+	log.Debugf("Reporting vertex %v failure to Mission Control", v)
+
+	// First, we'll add the failed vertex to our local prune view snapshot.
+	p.pruneViewSnapshot.vertexes[v] = struct{}{}
+
+	// With the vertex added, we'll now report back to the global prune
+	// view, with this new piece of information so it can be utilized for
+	// new payment sessions.
+	p.mc.Lock()
+	p.mc.failedVertexes[v] = time.Now()
+	p.mc.Unlock()
+}
+
+// ReportChannelFailure adds a channel to the graph prune view. The time the
+// channel was added is noted, as it'll be pruned from the global view after a
+// period of edgeDecay. However, the edge will remain pruned for the duration
+// of the *local* session. This ensures that we don't flap by continually
+// retrying an edge after its pruning has expired.
+//
+// TODO(roasbeef): also add value attempted to send and capacity of channel
+func (p *paymentSession) ReportEdgeFailure(e *edgeLocator) {
+	log.Debugf("Reporting edge %v failure to Mission Control", e)
+
+	// First, we'll add the failed edge to our local prune view snapshot.
+	p.pruneViewSnapshot.edges[*e] = struct{}{}
+
+	// With the edge added, we'll now report back to the global prune view,
+	// with this new piece of information so it can be utilized for new
+	// payment sessions.
+	p.mc.Lock()
```

Comment:
Not related to this PR, tho we should abstract this operation behind a method to prevent us from modifying mission control’s internal state directly and externally acquiring its mutexes. Same for the above method

Better yet, mission control itself deserves its own interface so that access to internal variables is prevented entirely

Reply:
Yes, the design of mission control definitely needs to be stepped up.